styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/TerminationHandler.java

```diff
@@ -96,7 +130,46 @@ private static boolean isMissingDependency(Optional<Integer> exitCode) {
     return exitCode.map(c -> c == MISSING_DEPS_EXIT_CODE).orElse(false);
   }
 
-  private static boolean shouldFailFast(Optional<Integer> exitCode) {
-    return exitCode.map(c -> c == FAIL_FAST_EXIT_CODE).orElse(false);
+  private boolean shouldFailFast(RunState state, Optional<Integer> exitCode) {
+    if (exitCode.isPresent() && exitCode.orElseThrow() == FAIL_FAST_EXIT_CODE) {
+      return true;
+    }
+
+    var workflow = workflows.get().get(state.workflowInstance().workflowId());
+    if (workflow == null) {
+      LOG.info("Workflow {} does not exist", state.workflowInstance().workflowId().toKey());
+      return true;
+    }
+
+    return workflow.configuration().retryCondition()
+        .map(s -> !retryConditionMet(state, exitCode, s))
+        .orElse(false);
+  }
+
+  @VisibleForTesting
+  boolean retryConditionMet(RunState state,
+                            Optional<Integer> exitCode,
+                            String retryCondition) {
+    var context = SimpleEvaluationContext.forReadOnlyDataBinding().build();
+    context.setVariable("exitCode", exitCode.orElse(null));
+    context.setVariable("tries", state.data().tries());
+    context.setVariable("triggerType", state.data().trigger().map(TriggerUtil::triggerType).orElse(null));
+    context.setVariable("consecutiveFailures", state.data().consecutiveFailures());
+
+    final Expression expression;
+    try {
+      expression = retryConditionExpressionCache.get(retryCondition,
+          () -> expressionParser.parseExpression(retryCondition));
+    } catch (ExecutionException | UncheckedExecutionException e) {
+      LOG.debug("Failed to parse retry condition `{}`", retryCondition, e.getCause());
+      return false;
+    }
+
+    try {
+      return TRUE.equals(expression.getValue(context, Boolean.class));
```

Comment:
`TRUE.equals` shouldn't be necessary, a Boolean.class type can be interpreted as a simple boolean already.

Reply:
Since `retryConditionMet` returns a raw value, unboxing a `Nullable Boolean` may result NPE. Return value of `expression.getValue` is `Nullable`.